examples/developing_estimators/sklearn_is_fitted.py

```diff
@@ -0,0 +1,154 @@
+`__sklearn_is_fitted__` method and `check_is_fitted` utility
+function as internal developer APIs. The `__sklearn_is_fitted__`
+method checks whether the estimator has been fitted by verifying
+the presence of the `_is_fitted` attribute.
+"""
+
+# %%
+# An example custom estimator implementing a simple classifier
+# ------------------------------------------------------------
+# This code snippet defines a custom estimator class called `CustomEstimator`
+# that extends the `BaseEstimator` and `ClassifierMixin` classes from
+# scikit-learn. It showcases the usage of the `__sklearn_is_fitted__` method
+# and the `check_is_fitted` utility function as internal developer APIs.
+
+# Author: Kushan <kushansharma1@gmail.com>
+#
+# License: BSD 3 clause
+
+from sklearn.base import BaseEstimator, ClassifierMixin
+from sklearn.utils.validation import check_is_fitted
+
+
+class CustomEstimator(BaseEstimator, ClassifierMixin):
+    def __init__(self, parameter=1):
+        self._is_fitted = False
+        self.parameter = parameter
+
+    def fit(self, X, y):
+        """
+        Fit the estimator to the training data.
+        Parameters:
+        -----------
+        - X : array-like of shape (n_samples, n_features)
+            The input samples.
+        - y : array-like of shape (n_samples,)
+            The target values.
+        Returns:
+        --------
+        - self : object
+            Returns self.
+        """
+        self.classes_ = sorted(set(y))
+        # Custom attribute to track if the estimator is fitted
+        self._is_fitted = True
+        return self
+
+    def predict(self, X):
+        """
+        Perform predictions using the fitted estimator.
+        Parameters:
+        -----------
+        - X : array-like of shape (n_samples, n_features)
+            The input samples.
+        Returns:
+        --------
+        - predictions : list
+            The predicted labels for the input samples.
+        Raises:
+        -------
+        - NotFittedError: If the estimator is not fitted.
+        """
+        check_is_fitted(self)
+        # Perform prediction logic
+        predictions = [self.classes_[0]] * len(X)
+        return predictions
+
+    def score(self, X, y):
+        """
+        Calculate the score of the estimator.
+        Parameters:
+        -----------
+        - X : array-like of shape (n_samples, n_features)
+            The input samples.
+        - y : array-like of shape (n_samples,)
+            The target values.
+        Returns:
+        --------
+        - score : float
+            The calculated score.
+        Raises:
+        -------
+        - NotFittedError: If the estimator is not fitted.
+        """
+        check_is_fitted(self)
+        # Perform scoring logic
+        return 0.5
+
+    def __sklearn_is_fitted__(self):
+        """Internal developer API. Returns True if the estimator is fitted,
+        False otherwise."""
```

Comment:
The first line of the docstring should be a single line. So maybe make it shorter, and add in following lines more explanation if needed.